File: app/resource_api_paths.py
"""Canonical resource API paths — single source for inventory list routes.

Primary list path: ``/resources/{canonical_type}`` (e.g. ``/resources/compute/disk``).
Legacy slug paths (``/resources/disks``) remain registered as aliases for compatibility.
"""

from __future__ import annotations

from app.inventory_standalone import is_standalone_inventory_type
from app.resource_page_registry import inventory_pages
LEGACY_SLUG_TO_CANONICAL: dict[str, str] = {
    "vms": "compute/vm",
    "disks": "compute/disk",
    "snapshots": "compute/snapshot",
    "aks": "containers/aks",
    "acr": "containers/acr",
    "storage": "storage/account",
    "publicips": "network/publicip",
    "vnets": "network/vnet",
    "nics": "network/nic",
    "natgateways": "network/nat",
    "loadbalancers": "network/loadbalancer",
    "appgateways": "network/appgateway",
    "nsgs": "network/nsg",
    "privateendpoints": "network/privateendpoint",
    "privatelinkservices": "network/privatelinkservice",
    "privatedns": "network/privatedns",
    "sql": "database/sql",
    "cosmosdb": "database/cosmosdb",
    "postgresql": "database/postgresql",
    "redis": "database/redis",
    "appservices": "appservice/webapp",
    "appserviceplans": "appservice/plan",
    "keyvaults": "security/keyvault",
}

def _merge_per_type_slugs() -> None:
    for _page in inventory_pages():
        if is_standalone_inventory_type(_page.canonical_type):
            LEGACY_SLUG_TO_CANONICAL.setdefault(_page.api_slug, _page.canonical_type)


_merge_per_type_slugs()

LEGACY_AGGREGATE_SLUGS: dict[str, tuple[str, ...]] = {
    "monitoring": ("monitoring/loganalytics", "monitoring/appinsights"),
    "integration": ("integration/apim", "integration/datafactory", "integration/logicapp"),
    "messaging": ("messaging/eventhub", "messaging/servicebus"),
    "analytics": (
        "analytics/databricks",
        "analytics/synapse",
        "analytics/adx",
        "analytics/mlworkspace",
    ),
    "backup": ("backup/recoveryvault",),
    "search": ("search/cognitivesearch",),
}


def canonical_api_path(canonical_type: str) -> str:
    ct = (canonical_type or "").strip().lower()
    return f"/resources/{ct}"


def legacy_slug_api_path(slug: str) -> str:
    return f"/resources/{(slug or '').strip().lower()}"


def _build_maps() -> tuple[dict[str, str], dict[str, str]]:
    canonical_to_path: dict[str, str] = {}
    path_to_canonical: dict[str, str] = {}

    for slug, ct in LEGACY_SLUG_TO_CANONICAL.items():
        legacy = legacy_slug_api_path(slug)
        canonical = canonical_api_path(ct)
        canonical_to_path[ct] = canonical
        path_to_canonical[legacy] = ct
        path_to_canonical[canonical] = ct

    for slug, types in LEGACY_AGGREGATE_SLUGS.items():
        path_to_canonical[legacy_slug_api_path(slug)] = types[0]  # primary for icon/sync hints

    return canonical_to_path, path_to_canonical


CANONICAL_TO_API_PATH, API_PATH_TO_CANONICAL = _build_maps()


def canonical_from_api_path(api_path: str) -> str | None:
    path = (api_path or "").strip().lower().rstrip("/")
    if not path.startswith("/"):
        path = f"/{path}"
    ct = API_PATH_TO_CANONICAL.get(path)
    if ct and not is_standalone_inventory_type(ct):
        return None
    return ct


def canonical_types_for_api_path(api_path: str) -> list[str]:
    path = (api_path or "").strip().lower().rstrip("/")
    if not path.startswith("/"):
        path = f"/{path}"
    slug = path.removeprefix("/resources/").strip("/")
    if slug in LEGACY_AGGREGATE_SLUGS:
        return [
            ct for ct in LEGACY_AGGREGATE_SLUGS[slug]
            if is_standalone_inventory_type(ct)
        ]
    ct = canonical_from_api_path(path)
    return [ct] if ct else []


def inventory_route_catalog() -> list[dict[str, str]]:
    rows: list[dict[str, str]] = []
    seen: set[str] = set()
    for slug, ct in sorted(LEGACY_SLUG_TO_CANONICAL.items(), key=lambda item: item[1]):
        if ct in seen or not is_standalone_inventory_type(ct):
            continue
        seen.add(ct)
        rows.append({
            "canonical_type": ct,
            "api_path": canonical_api_path(ct),
            "legacy_api_path": legacy_slug_api_path(slug),
            "legacy_slug": slug,
        })
    return rows

File: frontend/src/config/resourceApiPaths.js
/**
 * Canonical resource API paths — mirrors app/resource_api_paths.py.
 * Primary list path: /resources/{canonical_type} (e.g. /resources/compute/disk).
 */

import { PER_TYPE_API_PATH_TO_CANONICAL } from './resourcePageCatalog';

export function canonicalApiPath(canonicalType) {
  const ct = String(canonicalType || '').trim().toLowerCase();
  return `/resources/${ct}`;
}

/** Count key → canonical type (core inventory pages). */
export const COUNT_KEY_TO_CANONICAL = {
  vms: 'compute/vm',
  disks: 'compute/disk',
  snapshots: 'compute/snapshot',
  aks: 'containers/aks',
  acr: 'containers/acr',
  storage: 'storage/account',
  publicips: 'network/publicip',
  vnets: 'network/vnet',
  nics: 'network/nic',
  natgateways: 'network/nat',
  loadbalancers: 'network/loadbalancer',
  appgateways: 'network/appgateway',
  nsgs: 'network/nsg',
  privateendpoints: 'network/privateendpoint',
  privatelinkservices: 'network/privatelinkservice',
  privatedns: 'network/privatedns',
  sql: 'database/sql',
  cosmosdb: 'database/cosmosdb',
  postgresql: 'database/postgresql',
  redis: 'database/redis',
  appservices: 'appservice/webapp',
  appserviceplans: 'appservice/plan',
  keyvaults: 'security/keyvault',
  ...Object.fromEntries(
    Object.entries(PER_TYPE_API_PATH_TO_CANONICAL).map(([path, ct]) => [
      path.replace('/resources/', ''),
      ct,
    ]),
  ),
};

/** Legacy slug path → canonical type (includes canonical paths). */
export const API_PATH_TO_CANONICAL = (() => {
  const out = {};
  for (const [key, ct] of Object.entries(COUNT_KEY_TO_CANONICAL)) {
    out[`/resources/${key}`] = ct;
    out[canonicalApiPath(ct)] = ct;
  }
  return out;
})();

/** Canonical type → primary API list path. */
export const CANONICAL_TO_API_PATH = Object.fromEntries(
  Object.values(COUNT_KEY_TO_CANONICAL).map((ct) => [ct, canonicalApiPath(ct)]),
);

export function apiPathForCanonical(canonicalType) {
  const ct = String(canonicalType || '').trim().toLowerCase();
  return CANONICAL_TO_API_PATH[ct] || canonicalApiPath(ct);
}

export function apiPathForCountKey(countKey) {
  const key = String(countKey || '').trim().toLowerCase();
  const ct = COUNT_KEY_TO_CANONICAL[key];
  return ct ? apiPathForCanonical(ct) : null;
}

export function canonicalFromApiPath(apiPath) {
  const path = String(apiPath || '').trim().toLowerCase().replace(/\/+$/, '');
  return API_PATH_TO_CANONICAL[path] || null;
}

/** @deprecated Legacy aggregate list paths — resolve for scoped sync only. */
export const LEGACY_AGGREGATE_API_PATH_TO_TYPES = {
  '/resources/monitoring': ['monitoring/loganalytics', 'monitoring/appinsights'],
  '/resources/integration': ['integration/apim', 'integration/datafactory', 'integration/logicapp'],
  '/resources/messaging': ['messaging/eventhub', 'messaging/servicebus'],
  '/resources/analytics': [
    'analytics/databricks',
    'analytics/synapse',
    'analytics/adx',
    'analytics/mlworkspace',
  ],
  '/resources/backup': ['backup/recoveryvault'],
  '/resources/search': ['search/cognitivesearch'],
};

export function syncTypesForCanonicalApiPath(apiPath) {
  const path = String(apiPath || '').trim().toLowerCase().replace(/\/+$/, '');
  if (LEGACY_AGGREGATE_API_PATH_TO_TYPES[path]) {
    return [...LEGACY_AGGREGATE_API_PATH_TO_TYPES[path]];
  }
  const ct = canonicalFromApiPath(path);
  return ct ? [ct] : [];
}

/** Nested resource routes under canonical prefixes. */
export const CANONICAL_NESTED_PATHS = {
  vmSizing: (resourceGroup, vmName) =>
    `/resources/compute/vm/${encodeURIComponent(resourceGroup)}/${encodeURIComponent(vmName)}/sizing`,
  vmSizingOpenFinding: (resourceGroup, vmName) =>
    `/resources/compute/vm/${encodeURIComponent(resourceGroup)}/${encodeURIComponent(vmName)}/sizing/open-finding`,
  aksKubernetesVersions: '/resources/containers/aks/kubernetes-versions',
  aksPoolInstances: '/resources/containers/aks/pool-instances',
};
